### PR TITLE
CC-147:make HiveSchemaConverter handle logical types

### PR DESCRIPTION
**Pull request by liukrimhrim**

### CODE REVIEW

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 1]
```diff
@@ -90,6 +98,29 @@ public static TypeInfo convertMap(Schema schema) {
...
+      return TYPE_TO_TYPEINFO.get(schema.type());
+    }
+
+    //TODO: should take Hive version into consideration
+    switch (schema.name()) {
+      case Decimal.LOGICAL_NAME:
+        String scale = schema.parameters().get(Decimal.SCALE_FIELD);
+        String precision = schema.parameters().getOrDefault(CONNECT_AVRO_DECIMAL_PRECISION_PROP,
+            CONNECT_AVRO_DECIMAL_PRECISION_DEFAULT);
+        return new DecimalTypeInfo(Integer.parseInt(precision), Integer.parseInt(scale));
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: Is there any value for precision and/or scale that requires different type than `DecimalTypeInfo` ?

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 2]
```diff
@@ -26,11 +29,16 @@
...
 import java.util.List;
 import java.util.Map;
+import org.apache.kafka.connect.data.Time;
+import org.apache.kafka.connect.data.Timestamp;
 
 public class HiveSchemaConverter {
 
   private static final Map<Type, TypeInfo> TYPE_TO_TYPEINFO;
 
+  private static final String CONNECT_AVRO_DECIMAL_PRECISION_PROP = "connect.decimal.precision";
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: Any new configuration options that need to be introduced here, will need to be defined along with the other ones for Hive in:
https://github.com/confluentinc/kafka-connect-storage-common/blob/master/hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 3]
```diff
@@ -90,6 +98,29 @@ public static TypeInfo convertMap(Schema schema) {
   }
 
   public static TypeInfo convertPrimitive(Schema schema) {
-    return TYPE_TO_TYPEINFO.get(schema.type());
+    if (schema.name() == null) {
+      return TYPE_TO_TYPEINFO.get(schema.type());
+    }
+
+    //TODO: should take Hive version into consideration
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: See comment about defining new config properties above, that will allow you to enable or disable this feature

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 4]
```diff
@@ -90,6 +98,29 @@ public static TypeInfo convertMap(Schema schema) {
...
+    switch (schema.name()) {
+      case Decimal.LOGICAL_NAME:
+        String scale = schema.parameters().get(Decimal.SCALE_FIELD);
+        String precision = schema.parameters().getOrDefault(CONNECT_AVRO_DECIMAL_PRECISION_PROP,
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: In other config properties you'll see examples of getting the default from a config that has a default defined.

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java (hunk 1 above)
Comment (liukrimhrim) [reply hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: You mean "is there any type other than `DecimalTypeInfo` that requires precision and scale"? No. This is the only logical type that requires it.

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java (hunk 4 above)
Comment (liukrimhrim) [reply hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: A note on the default value of precision: `AvroData` uses 64 as default, but Hive allows at most 38, so I use 38 as default.

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java (hunk 2 above)
Comment (liukrimhrim) [reply hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: This property is name is defined in `AvroData`. Here, I'm just using it to retrieve the value of precision, which was set when the runtime connect record was created using `AvroData`. I think users of Hive connector should be given the option to config this name.

**Comment by liukrimhrim** [comment]
> Thanks @liukrimhrim !
> I added a few comments regarding config definition.
> I noticed that [confluentinc/kafka-connect-hdfs#428](https://github.com/confluentinc/kafka-connect-hdfs/pull/428)
> has added tests for this code, but I also wonder if we could some self-contained unit tests here too?

Currently, there is no test folder yet. Should I add the test folder manually or if there is any way to set up the test folder automatically?

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java [hunk 5]
```diff
@@ -69,6 +69,25 @@
...
+  public static final String HIVE_LOGICAL_TYPES_ENABLE_CONFIG = "hive.logical.types.enable";
+  public static final String HIVE_LOGICAL_TYPES_ENABLE_DOC = "Whether enable hive logical types:"
+      + "DECIMAL, DATE, INTERVAL, TIMESTAMP."
+      + "Please be noted that some of the logical types are not available "
+      + "in earlier versions of Hive. Better to disable hive logical type"
+      + "if using these earlier versions";
+  public static final boolean HIVE_LOGICAL_TYPES_ENABLE_DEFAULT = false;
+  public static final String HIVE_LOGICAL_TYPES_ENABLE_DISPLAY = "Enable Hive Logical Types";
+
+  public static final String DECIMAL_PRECISION_CAP_CONFIG = "decimal.precision.cap";
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java]: let's not use the word `cap`. It's clear that this refers to an upper limit

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java [hunk 6]
```diff
@@ -69,6 +69,25 @@
...
+  public static final String DECIMAL_PRECISION_CAP_DOC = "Runtime Connect records with precision"
+      + "greater than the cap will be written to sink as bytes rather than hive logical type"
+      + "DECIMAL. Hive allows at most 38 digits precision, "
+      + "if this config field is set to values greater than 38, "
+      + "the default value 38 will be used as the cap.";
+  public static final int DECIMAL_PRECISION_CAP_DEFAULT = 38;
+  public static final String DECIMAL_PRECISION_CAP_DISPLAY = "Decimal Precision Cap";
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java]: no cap as above

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java [hunk 7]
```diff
@@ -69,6 +69,25 @@
...
+      + "if using these earlier versions";
+  public static final boolean HIVE_LOGICAL_TYPES_ENABLE_DEFAULT = false;
+  public static final String HIVE_LOGICAL_TYPES_ENABLE_DISPLAY = "Enable Hive Logical Types";
+
+  public static final String DECIMAL_PRECISION_CAP_CONFIG = "decimal.precision.cap";
+  public static final String DECIMAL_PRECISION_CAP_DOC = "Runtime Connect records with precision"
+      + "greater than the cap will be written to sink as bytes rather than hive logical type"
+      + "DECIMAL. Hive allows at most 38 digits precision, "
+      + "if this config field is set to values greater than 38, "
+      + "the default value 38 will be used as the cap.";
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java]: `... will be used instead`

Comment (kkonstantine) [reply hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java]: why not throw an error with a validator? Just asking

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java [hunk 8]
```diff
@@ -175,6 +194,37 @@
...
+          ++orderInGroup,
+          Width.SHORT,
+          HIVE_LOGICAL_TYPES_ENABLE_DISPLAY
+      );
+
+      CONFIG_DEF.define(
+          DECIMAL_PRECISION_CAP_CONFIG,
+          Type.INT,
+          DECIMAL_PRECISION_CAP_DEFAULT,
+          Importance.MEDIUM,
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveConfig.java]: as mentioned above, a validator and a recommender might be a good option here. 
Feel free to create a single inner static class that implements both interfaces.

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 9]
```diff
@@ -26,11 +29,21 @@
...
+import org.apache.kafka.connect.data.Timestamp;
 
 public class HiveSchemaConverter {
 
   private static final Map<Type, TypeInfo> TYPE_TO_TYPEINFO;
 
+  // the name here is consistent with io.confluent.connect.avro.AvroData, when connect Decimal
+  // schema is created, this property name is used to set precision.
+  // We have to use the exact name to retrieve precision value.
+  static final String CONNECT_AVRO_DECIMAL_PRECISION_PROP = "connect.decimal.precision";
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: unfortunate that this is not public in `AvroData` but instead just package private. Let's not repeat the same mistake here.

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 10]
```diff
@@ -26,11 +29,21 @@
...
+
+  // this is the maximum digits Hive allows for DECIMAL type.
+  static final int DECIMAL_PRECISION_DEFAULT = 38;
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: public as above

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java [hunk 11]
```diff
@@ -55,6 +68,17 @@
     return columns;
   }
 
+  public static List<FieldSchema>  convertSchemaMaybeLogical(Schema schema) {
```
Comment (kkonstantine) [inline hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: nit: two space between the type and the method name. Surprised that checkstyle doesn't mind

File: hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java (hunk 2 above)
Comment (kkonstantine) [reply hive/src/main/java/io/confluent/connect/storage/hive/HiveSchemaConverter.java]: ok.

**Comment by liukrimhrim** [comment]
Closing this old PR . Link to the new one: https://github.com/confluentinc/kafka-connect-storage-common/pull/114

**Pull request closed without merging**